flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/logical/PushPartitionIntoTableSourceScanRule.scala

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.logical
+
+import org.apache.flink.table.calcite.{FlinkContext, FlinkTypeFactory}
+import org.apache.flink.table.plan.schema.{FlinkRelOptTable, TableSourceTable}
+import org.apache.flink.table.plan.stats.FlinkStatistic
+import org.apache.flink.table.plan.util.{FlinkRelOptUtil, PartitionPruner, RexNodeExtractor}
+import org.apache.flink.table.sources.PartitionableTableSource
+
+import org.apache.calcite.plan.RelOptRule.{none, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rel.core.Filter
+import org.apache.calcite.rel.logical.LogicalTableScan
+import org.apache.calcite.rex.{RexInputRef, RexNode, RexShuttle}
+
+import scala.collection.JavaConversions._
+
+/**
+  * Planner rule that tries to push partitions evaluated by filter condition into a
+  * [[PartitionableTableSource]].
+  */
+class PushPartitionIntoTableSourceScanRule extends RelOptRule(
+  operand(classOf[Filter],
+    operand(classOf[LogicalTableScan], none)),
+  "PushPartitionIntoTableSourceScanRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val filter: Filter = call.rel(0)
+    if (filter.getCondition == null) {
+      return false
+    }
+
+    val scan: LogicalTableScan = call.rel(1)
+    scan.getTable.unwrap(classOf[TableSourceTable[_]]) match {
+      case table: TableSourceTable[_] =>
+        table.tableSource match {
+          case p: PartitionableTableSource => p.getPartitionFieldNames.nonEmpty
+          case _ => false
+        }
+      case _ => false
+    }
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val filter: Filter = call.rel(0)
+    val scan: LogicalTableScan = call.rel(1)
+    val table: FlinkRelOptTable = scan.getTable.asInstanceOf[FlinkRelOptTable]
+    pushPartitionIntoScan(call, filter, scan, table)
+  }
+
+  private def pushPartitionIntoScan(
+      call: RelOptRuleCall,
+      filter: Filter,
+      scan: LogicalTableScan,
+      relOptTable: FlinkRelOptTable): Unit = {
+
+    val tableSourceTable = relOptTable.unwrap(classOf[TableSourceTable[_]])
+    val tableSource = tableSourceTable.tableSource.asInstanceOf[PartitionableTableSource]
+    val partitionFieldNames = tableSource.getPartitionFieldNames.toList.toArray
+    val inputFieldType = filter.getInput.getRowType
+
+    val relBuilder = call.builder()
+    val maxCnfNodeCount = FlinkRelOptUtil.getMaxCnfNodeCount(scan)
+    val (partitionPredicate, nonPartitionPredicate) =
+      RexNodeExtractor.extractPartitionPredicates(
+        filter.getCondition,
+        maxCnfNodeCount,
+        inputFieldType.getFieldNames.toList.toArray,
+        relBuilder.getRexBuilder,
+        partitionFieldNames
+      )
+
+    if (partitionPredicate.isAlwaysTrue) {
```

Comment:
If this happens, will optimizer ended up in infinite loop?

Reply:
no, that means there is no partition predicates in filter, the rule will not be matched for this filter later.

Reply:
There is no equivalent nodes generated, so  infinite loop will not happen in this case.